src/libraries/System.Private.CoreLib/src/System/TimeSpan.cs

```diff
@@ -308,6 +308,149 @@ public TimeSpan Duration()
 
         public override int GetHashCode() => _ticks.GetHashCode();
 
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TimeSpan"/> structure to a specified number of
+        /// days.
+        /// </summary>
+        /// <param name="days">Number of days.</param>
+        /// <returns></returns>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// The parameters specify a <see cref="TimeSpan"/> value less than <see cref="MinValue"/> or greater than <see cref="MaxValue"/>
+        /// </exception>
+        public static TimeSpan FromDays(int days) => FromDays(days, 0);
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TimeSpan"/> structure to a specified number of
+        /// days, hours, minutes, seconds, milliseconds, and microseconds.
+        /// </summary>
+        /// <param name="days">Number of days.</param>
+        /// <param name="hours">Number of hours.</param>
+        /// <param name="minutes">Number of minutes.</param>
+        /// <param name="seconds">Number of seconds.</param>
+        /// <param name="milliseconds">Number of milliseconds.</param>
+        /// <param name="microseconds">Number of microseconds.</param>
+        /// <returns></returns>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// The parameters specify a <see cref="TimeSpan"/> value less than <see cref="MinValue"/> or greater than <see cref="MaxValue"/>
+        /// </exception>
+        public static TimeSpan FromDays(int days, int hours = 0, long minutes = 0, long seconds = 0, long milliseconds = 0, long microseconds = 0)
+        {
+            Int128 totalMicroseconds = ((Int128)days * (Int128)MicrosecondsPerDay)
```

Comment:
This is definitely the simplest way to do this, but its also potentially more expensive than necessary for the common cases.

For a non-overflow case, what's the perf look like compared to doing (with today's `double` overloads) `FromDays() + FromHours() + FromMinutes() + FromSeconds() + ...`

Reply:
> more expensive than necessary for the common cases.

Definitively, but its the only way I could see to allow the full range of inputs without overflowing (before the check of final result).
I believe the other 2 options would be to limit the range of inputs such that they are not able to overflow an `Int64`, or either accepting or checking for overflow.

The single argument overloads (like `FromDays(int)`) is, of course, trivial to optimize.

Unless I missed something it does not look like a chain of `FromDays(double) + ..` is the competitor though.
Its a bit quick and dirty (copy paste `TimeSpan.cs` and making it compile), but I think this should be right.
https://github.com/tommysor/TimeSpanMathPerf
* Doubles: `FromDays(2.0) + .. + FromMicroseconds(7.0)`
* Int128: `FromDays(2, 3, 4, 5, 6, 7)` (Implementation from this PR)
* Int64: `new TimeSpan(2, 3, 4, 5, 6, 7)` (same logic as Int128, but without the cast to `Int128`)
```
BenchmarkDotNet v0.13.12, Debian GNU/Linux 12 (bookworm) (container)
AMD EPYC 7763, 1 CPU, 2 logical cores and 1 physical core
.NET SDK 8.0.101
  [Host]     : .NET 8.0.1 (8.0.123.58001), X64 RyuJIT AVX2
  DefaultJob : .NET 8.0.1 (8.0.123.58001), X64 RyuJIT AVX2
```
| Method  | Mean       | Error     | StdDev    | Median     | Ratio | RatioSD |
|-------- |-----------:|----------:|----------:|-----------:|------:|--------:|
| Doubles | 13.8305 ns | 0.4129 ns | 1.2045 ns | 13.4968 ns | 1.000 |    0.00 |
| Int128  |  6.9789 ns | 0.1914 ns | 0.5553 ns |  6.8646 ns | 0.508 |    0.06 |
| Int64   |  0.0283 ns | 0.0198 ns | 0.0478 ns |  0.0000 ns | 0.002 |    0.00 |

Reply:
> new TimeSpan(2, 3, 4, 5, 6, 7)

Note that you may want to test this with something the JIT can't see as constant. It's fully folding it away and so isn't a great comparison.

---------

The basic here is that we're already 2x as fast as what users can get today (if they want to guarantee accuracy) via `double`, but we're likely still quite a bit slower than we could be and the current code can't be constant folded. The multiplications that `Int128` is doing are also more expensive than needed by quite a bit since we know both inputs are within 64-bits.

We can't really do much without the JIT recognizing `BigMul` as constant foldable. However, you can likely make it a little cheaper with the following:
```csharp
public static TimeSpan FromDays(int days, int hours = 0, long minutes = 0, long seconds = 0, long milliseconds = 0, long microseconds = 0)
{
    Int128 totalMicroseconds = Math.BigMul(days, MicrosecondsPerDay)
                             + Math.BigMul(hours, MicrosecondsPerHour)
                             + Math.BigMul(minutes, MicrosecondsPerMinute)
                             + Math.BigMul(seconds, MicrosecondsPerSecond)
                             + Math.BigMul(milliseconds, MicrosecondsPerMillisecond)
                             + microseconds;

    long lowerMicroseconds = totalMicroseconds.Lower;
    long upperMicroseconds = totalMicroseconds.Upper;

    if (((upperMicroseconds == 0) && (lowerMicroseconds > MaxMicroseconds)) || (upperMicroseconds != -1) || (totalMicroseconds < MinMicroseconds))
    {
        ThrowHelper.ThrowArgumentOutOfRange_TimeSpanTooLong();
    }
    var ticks = (long)totalMicroseconds * TicksPerMicrosecond;
    return TimeSpan.FromTicks(ticks);
}

// In Math.cs
internal static Int128 BigMul(long a, long b)
{
    long upper = BigMul(a, b, out long lower);
    return new Int128((ulong)upper, (ulong)lower);
}
```

This should do less work for the multiplications and keeps it very cheap for the additions. I expect doing checked addition at each step would be overall more expensive than what `Int128` is doing and we know it can't overflow `Int128` already. We can then make it cheaper over time by recognizing a couple interesting patterns in the JIT (ones we want to recognize long term anyways).

Reply:
> (upperMicroseconds != -1)

@tannergooding Is this for checking the negative sign?

Reply:
Correct, which is allowed for TimeSpan, just only if its within `[MinMicroseconds, MaxMicroseconds]`

The `upperMicroseconds == 0` means that its a small positive (fits into 64-bits). If its a small positive and greater than `MaxMicroseconds`, we should throw. Here we check by comparing `lowerMicroseconds` as `unsigned` since we want the most significant bit being set to throw.

The `upperMicrosedconds != -1` then means its either a large positive or a large negative value (takes more than 64-bits), in which case we should throw.

Otherwise it's a small negative (fits into 64-bits) and we should throw if the lower Microseconds is less than MinMicroseconds. In this case, we again check by comparing the `lowerMicroseconds` as `unsigned`, because we're ultimately wanting it to be between `[-1, MinMicroseconds]` which in hex is `[0xFFFF_FFFF_FFFF_FFFF, 0xF333_3333_3333_3334]`, so we want the most significant bit set and the value to be greater than or equal to `MinMicroseconds` in unsigned form

Reply:
> fully folding it away

wops, no wonder the Int64 was so ridiculously fast.

And yes, seems like BigMul is a little cheaper (not tested the change to overflow check since I don't have access to internal `totalMicroseconds.Lower` in that benchmark repo)

https://github.com/tommysor/TimeSpanMathPerf
* Doubles: `FromDays(_daysDouble) + .. + FromMicroseconds(_microsecondsDouble)`
* Int128: `FromDays(_daysInt, .., .., .., .., _microsecondsInt)` (Implementation from this PR)
* Int128_BigMul: Using `BigMul` for multiplication
  Still same overflow check since my copied code does not have access to the internal `totalMicroseconds.Lower`
* Int64: new `TimeSpan(_daysInt, .., .., .., .., _microsecondsInt)` (same logic as Int128, but without the cast to Int128)
```
BenchmarkDotNet v0.13.12, Debian GNU/Linux 12 (bookworm) (container)
AMD EPYC 7763, 1 CPU, 2 logical cores and 1 physical core
.NET SDK 8.0.101
  [Host]     : .NET 8.0.1 (8.0.123.58001), X64 RyuJIT AVX2
  DefaultJob : .NET 8.0.1 (8.0.123.58001), X64 RyuJIT AVX2
```
| Method                  | Mean       | Error     | StdDev    | Median     | Ratio | RatioSD |
|------------------------ |-----------:|----------:|----------:|-----------:|------:|--------:|
| Doubles                 | 15.9139 ns | 0.3832 ns | 1.0871 ns | 15.4181 ns |  1.00 |    0.00 |
| Int128                  |  7.3156 ns | 0.1990 ns | 0.5835 ns |  7.0550 ns |  0.46 |    0.05 |
| Int128_BigMul           |  6.4639 ns | 0.2071 ns | 0.6041 ns |  6.2065 ns |  0.41 |    0.05 |
| Int64                   |  1.7092 ns | 0.0677 ns | 0.1698 ns |  1.6955 ns |  0.11 |    0.01 |

Reply:
Looks BigMul is the way to go. I am not concerning about the overflowing case performance as I am not expecting this will occur much with this APIs. @tommysor thanks for getting this numbers. Could you please update the code so we can move forward with this PR?